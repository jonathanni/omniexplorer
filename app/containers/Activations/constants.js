/*
 *
 * Activations constants
 *
 */

export const LOAD_ACTIVATIONS = 'app/Activations/LOAD_ACTIVATIONS';
export const LOAD_ACTIVATIONS_SUCCESS =
  'app/Activations/LOAD_ACTIVATIONS_SUCCESS';
