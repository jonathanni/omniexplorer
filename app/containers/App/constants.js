/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */
import generateTemplate from 'utils/generateTemplate';

export const API_URL_BASE = 'https://api.omniexplorer.info/v1';
export const API_URL_BLOCKCHAIN_BTC_BALANCE =
  'https://blockchain.info/balance?cors=true&active=';
export const FN_API_URL_BLOCKCHAIR_BTC_BALANCE = (data) =>
  generateTemplate`https://api.blockchair.com/bitcoin/dashboards/address/${'address'}?state=latest&limit=0,0`(
    data,
  );
export const DEFAULT_LOCALE = 'en';
export const DEFAULT_NOT_NUMBER = '---';
export const ECOSYSTEM_PROD = 1;
export const ECOSYSTEM_TEST = 2;
export const ECOSYSTEM_PROD_NAME = 'Production';
export const ECOSYSTEM_TEST_NAME = 'Test';
export const FIRST_BLOCK = 252317;
export const FEATURE_ACTIVATION_TYPE_INT = 65534;
