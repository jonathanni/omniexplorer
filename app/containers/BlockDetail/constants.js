/*
 *
 * BlockDetail constants
 *
 */

export const LOAD_BLOCK = 'omniexplorer/App/LOAD_BLOCK';
export const LOAD_BLOCK_SUCCESS = 'omniexplorer/App/LOAD_BLOCK_SUCCESS';
export const ALL_BLOCK_TRANSACTIONS = 'omniexplorer/App/ALL_BLOCK_TRANSACTIONS';
export const VALID_BLOCK_TRANSACTIONS =
  'omniexplorer/App/VALID_BLOCK_TRANSACTIONS';
export const INVALID_BLOCK_TRANSACTIONS =
  'omniexplorer/App/INVALID_BLOCK_TRANSACTIONS';
