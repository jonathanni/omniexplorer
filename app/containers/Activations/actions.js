/*
 *
 * Activations actions
 *
 */

import { LOAD_ACTIVATIONS, LOAD_ACTIVATIONS_SUCCESS } from './constants';

/**
 * Load activations, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_ACTIVATIONS
 */
export function loadActivations() {
  return {
    type: LOAD_ACTIVATIONS,
  };
}

/**
 * Dispatched when the activations are loaded by the request saga
 *
 * @param  {array} Activations The activations data
 *
 * @return {object} An action object with a type of LOAD_ACTIVATIONS_SUCCESS passing the transactions
 */
export function activationsLoaded(activations) {
  return {
    type: LOAD_ACTIVATIONS_SUCCESS,
    activations,
  };
}
