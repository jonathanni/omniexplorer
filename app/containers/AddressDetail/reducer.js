/*
 *
 * AddressDetail reducer
 *
 */
import produce from 'immer';
import { LOAD_ADDRESS, LOAD_ADDRESS_SUCCESS } from './constants';

const initialAddress = {
  balance: [],
};

export const initialState = {
  loading: false,
  address: initialAddress,
};

/* eslint-disable default-case, no-param-reassign */
const addressDetailReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case LOAD_ADDRESS:
        draft.loading = true;
        draft.address = initialAddress;
        break;
      case LOAD_ADDRESS_SUCCESS:
        draft.loading = false;
        draft.address = action.address;
        break;
    }
  });

export default addressDetailReducer;
