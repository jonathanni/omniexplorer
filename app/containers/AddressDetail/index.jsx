/**
 *
 * AddressDetail
 *
 */

import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { useParams } from 'react-router-dom';

import styled from 'styled-components';
import { Col, Container, Row } from 'reactstrap';

import Transactions from 'containers/Transactions';
import Wallet from 'components/Wallet';

import { useInjectSaga } from 'utils/injectSaga';
import { useInjectReducer } from 'utils/injectReducer';

import makeSelectAddressDetail from './selectors';
import reducer from './reducer';
import { loadAddress } from './actions';
import sagaAddress from './saga';

const Layout = styled(Container)`
  background-color: white;
  padding: 0;
`;

export function AddressDetail(props) {
  const { address } = useParams();

  useInjectReducer({
    key: 'addressDetail',
    reducer,
  });
  useInjectSaga({
    key: 'addressDetail',
    saga: sagaAddress,
  });

  useEffect(() => {
    props.loadAddress(address);
  }, [address]);

  if (props.loading) {
    return;
  }

  return (
    <Layout fluid>
      <Row>
        <Col sm>
          <Wallet {...props.addressdetail} addr={address} />
        </Col>
      </Row>
      <Row>
        <Col sm>
          <Transactions addr={address} {...props} currentPage={1} />
        </Col>
      </Row>
    </Layout>
  );
}

AddressDetail.propTypes = {
  dispatch: PropTypes.func.isRequired,
  loadAddress: PropTypes.func,
  addressdetail: PropTypes.object.isRequired,
  match: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  addressdetail: makeSelectAddressDetail(),
});

function mapDispatchToProps(dispatch) {
  return {
    loadAddress: addr => dispatch(loadAddress(addr)),
    dispatch,
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
)(AddressDetail);
