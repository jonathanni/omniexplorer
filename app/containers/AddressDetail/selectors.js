import { createSelector } from 'reselect';
import { initialState } from './reducer';

/**
 * Direct selector to the addressDetail state domain
 */
const selectAddressDetailDomain = (state) =>
  state.addressDetail || initialState;

/**
 * Other specific selectors
 */

/**
 * Default selector used by AddressDetail
 */

const makeSelectAddressDetail = () =>
  createSelector(selectAddressDetailDomain, (substate) => substate);

export default makeSelectAddressDetail;
export { selectAddressDetailDomain };
