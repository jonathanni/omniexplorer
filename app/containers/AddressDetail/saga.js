import { all, call, put, take } from 'redux-saga/effects';
import { LOAD_ADDRESS } from 'containers/AddressDetail/constants';
import {
  API_URL_BASE,
  API_URL_BLOCKCHAIN_BTC_BALANCE,
  FN_API_URL_BLOCKCHAIR_BTC_BALANCE,
} from 'containers/App/constants';
import { updateFetch } from 'components/Token/actions';
import { addressLoaded } from 'containers/AddressDetail/actions';
import encoderURIParams from 'utils/encoderURIParams';

import request from 'utils/request';
import isNil from 'lodash/isNil';

export function* getAddress({ addr }) {
  const requestURL = `${API_URL_BASE}/address/addr`;

  const body = encoderURIParams({ addr });
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body,
  };

  /* eslint-disable prefer-const */
  // get BTC balance from blockchain.info for the given wallet
  const urlBTCBalance = `${API_URL_BLOCKCHAIN_BTC_BALANCE}${addr}`;
  let [wallet, btcBalance] = yield all([
    call(request, requestURL, options),
    call(request, urlBTCBalance),
  ]);

  /**
   // use btc balance from blockchain.info response
   const btcBalanceValue = btcBalance[addr].final_balance;
   const walletBTCBalance = wallet.balance.find(x => x.id == 0);
   if (walletBTCBalance) walletBTCBalance.value = btcBalanceValue;
   * */

  let btcBalanceValue;

  // if there is a valid response use btc balance from blockchain.info response
  if (btcBalance[addr] && !isNil(btcBalance[addr].final_balance)) {
    btcBalanceValue = btcBalance[addr].final_balance;
  } else {
    // if blockchain.info retrieves an error try with blockchair
    const urlBTCBalanceAlternative = FN_API_URL_BLOCKCHAIR_BTC_BALANCE({
      address: addr,
    });
    btcBalance = yield call(request, urlBTCBalanceAlternative);
    // use btc balance from blockchair.com response
    btcBalanceValue = btcBalance.data[addr].address.balance;
  }

  const walletBTCBalance = wallet.balance.find((x) => x.id === 0);
  if (walletBTCBalance) walletBTCBalance.value = btcBalanceValue;

  yield put(addressLoaded(wallet));
  yield wallet.balance.map((property) =>
    put(updateFetch(property.propertyinfo)),
  );
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* root() {
  while (true) {
    const payload = yield take(LOAD_ADDRESS);
    yield call(getAddress, payload);
  }
}
