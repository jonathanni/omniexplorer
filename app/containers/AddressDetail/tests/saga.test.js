/**
 * Tests for AddressDetail sagas
 */
import {
  API_URL_BASE,
  API_URL_BLOCKCHAIN_BTC_BALANCE,
  FN_API_URL_BLOCKCHAIR_BTC_BALANCE,
} from 'containers/App/constants';

import { all, put, take, call } from 'redux-saga/effects';
import { cloneableGenerator } from '@redux-saga/testing-utils';
import request from 'utils/request';
import encoderURIParams from 'utils/encoderURIParams';

import { updateFetch } from 'components/Token/actions';
import { addressLoaded } from 'containers/AddressDetail/actions';

import { LOAD_ADDRESS } from 'containers/AddressDetail/constants';
import root, { getAddress } from '../saga';

const addr = '1FeexV6bAHb8ybZjqQMjJrcCrHGW9sb6uF';

/* eslint-disable redux-saga/yield-effects */
describe('getAddress Saga', () => {
  let getAddressGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getAddressGenerator = getAddress({ addr });

    const selectDescriptor = getAddressGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getAddressGenerator.next(addr).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  const saga = cloneableGenerator(getAddress)({ addr });
  const response = {
    balance: [
      {
        divisible: true,
        frozen: '0',
        id: '0',
        pendingneg: '0',
        pendingpos: '0',
        propertyinfo: {
          blocktime: 1231006505,
          data:
            'The Times 03/Jan/2009 Chancellor on brink of second bailout for banks',
          divisible: true,
          flags: {},
          issuer: 'Satoshi Nakamoto',
          name: 'BTC',
          propertyid: 0,
          rdata: null,
          registered: false,
          totaltokens: '18637850.00000000',
          url: 'http://www.bitcoin.org',
        },
        symbol: 'BTC',
        value: '0',
      },
    ],
  };

  const responseBalance = {
    [addr]: {
      final_balance: 7995720580428,
      n_tx: 368,
      total_received: 7995720580428,
    },
  };

  const responseBalanceBlockchair = {
    data: {
      [addr]: {
        address: {
          balance: 7995720579651,
          received: 7995720579651,
          spent: 0,
          output_count: 368,
        },
      },
    },
  };

  const url = `${API_URL_BASE}/address/addr`;
  const urlBTCBalance = `${API_URL_BLOCKCHAIN_BTC_BALANCE}${addr}`;
  const body = encoderURIParams({ addr });
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body,
  };

  it('should dispatch the addressLoaded action if it requests the data successfully', () => {
    expect(saga.next().value).toEqual(
      all([call(request, url, options), call(request, urlBTCBalance)]),
    );
  });

  describe('and blockchain.info provides a valid response', () => {
    let clone;

    beforeAll(() => {
      clone = saga.clone();
    });

    it('does provide a valid response', () => {
      expect(clone.next([response, responseBalance]).value).toEqual(
        put(addressLoaded(response)),
      );
      expect(clone.next().value).toEqual([
        put(updateFetch(response.balance[0].propertyinfo)),
      ]);
    });
  });

  describe('and blockchain.info did not provide a valid response', () => {
    let clone;

    beforeAll(() => {
      clone = saga.clone();
    });

    it('requests from blockchair and provides a valid response', () => {
      const urlBTCBalanceAlternative = FN_API_URL_BLOCKCHAIR_BTC_BALANCE({
        address: addr,
      });

      expect(clone.next([response, {}]).value).toEqual(
        call(request, urlBTCBalanceAlternative),
      );
      expect(clone.next(responseBalanceBlockchair).value).toEqual(
        put(addressLoaded(response)),
      );
      expect(clone.next().value).toEqual([
        put(updateFetch(response.balance[0].propertyinfo)),
      ]);
    });
  });

  /*
  it('should call the addressLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getAddressGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(addressLoadingError(response)));
  });
  */
});

describe('Address detail Saga', () => {
  it('should start task to watch for LOAD_ADDRESS action', () => {
    // arrange
    const rootSaga = root();
    const payload = { addr };

    // assert
    expect(rootSaga.next().value).toEqual(take(LOAD_ADDRESS));
    expect(rootSaga.next(payload).value).toEqual(call(getAddress, payload));
  });
});
