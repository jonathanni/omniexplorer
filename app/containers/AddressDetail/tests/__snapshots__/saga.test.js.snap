// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAddress Saga should dispatch the addressLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": true,
  "payload": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://api.omniexplorer.info/v1/address/addr",
          Object {
            "body": "addr=17ScKNXo4cL8DyfWfcCWu1uJySQuJm7iKx",
            "headers": Object {
              "Content-Type": "application/x-www-form-urlencoded",
            },
            "method": "POST",
          },
        ],
        "context": null,
        "fn": [Function],
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://blockchain.info/balance?cors=true&active=17ScKNXo4cL8DyfWfcCWu1uJySQuJm7iKx",
        ],
        "context": null,
        "fn": [Function],
      },
      "type": "CALL",
    },
  ],
  "type": "ALL",
}
`;

exports[`getAddress Saga should dispatch the addressLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.blockchair.com/bitcoin/dashboards/address/17ScKNXo4cL8DyfWfcCWu1uJySQuJm7iKx?state=latest&limit=0,0",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
